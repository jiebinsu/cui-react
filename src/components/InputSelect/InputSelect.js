import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";

const InputSelect = ({
  id,
  name,
  className,
  label,
  hint,
  hasError,
  errorMsg,
  options,
  placeholder,
  width,
  ...props
}) => {
  const classes = classNames("coop-form__field coop-form__select", [
    hasError && "coop-form__invalid",
    width && `coop-form__input--width-${width}`,
    className,
  ]);

  return (
    <div className="coop-form__row">
      {label && (
        <label className="coop-form__label" htmlFor={id}>
          {label}
        </label>
      )}
      {hint && (
        <p id={`${id}-hint`} className="coop-form__hint">
          This is an example hint
        </p>
      )}
      {hasError && (
        <p id={`${id}-error`} className="coop-form__error">
          {errorMsg}
        </p>
      )}
      <select
        id={id}
        name={name}
        className={classes}
        aria-describedby={`${id}-hint`}
        {...props}
      >
        {placeholder && (
          <option value="" disabled selected hidden>
            {placeholder}
          </option>
        )}
        {options.map(({ value, text }) => (
          <option key={value} value={value}>
            {text}
          </option>
        ))}
      </select>
    </div>
  );
};

InputSelect.defaultProps = {
  className: null,
  label: null,
  hint: null,
  hasError: false,
  errorMsg: null,
  options: [],
  placeholder: null,
  width: null,
};

InputSelect.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  label: PropTypes.string,
  hint: PropTypes.string,
  hasError: PropTypes.bool,
  errorMsg: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string || PropTypes.number,
      value: PropTypes.string || PropTypes.number,
    })
  ),
  placeholder: PropTypes.string,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default InputSelect;
