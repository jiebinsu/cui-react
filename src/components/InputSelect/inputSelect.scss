@import "@coopdigital/foundations-forms/dist/forms.css";
