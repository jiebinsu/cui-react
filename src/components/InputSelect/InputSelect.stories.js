import React from "react";
import InputSelect from "./InputSelect";

export default {
  title: "Base/InputSelect",
  component: InputSelect,
};

export const Basic = (args) => <InputSelect {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  options: [
    {
      value: "1",
      text: "one",
    },
    {
      value: "2",
      text: "two",
    },
  ],
};

export const All = () => {
  const options = [
    {
      value: "1",
      text: "one",
    },
    {
      value: "2",
      text: "two",
    },
  ];
  return (
    <>
      <InputSelect id="id" name="name" options={options} />
      <InputSelect
        id="id"
        name="name"
        options={options}
        placeholder="Placeholder"
      />
      <InputSelect id="id" name="name" label="With Label" options={options} />
      <InputSelect id="id" name="name" hint="With Hint" options={options} />
      <InputSelect
        id="id"
        name="name"
        label="With Label and Hint"
        hint="With Label and Hint"
        options={options}
      />
    </>
  );
};

export const WithError = () => {
  const options = [
    {
      value: "1",
      text: "one",
    },
    {
      value: "2",
      text: "two",
    },
  ];
  return (
    <>
      <InputSelect
        id="id"
        name="name"
        hasError
        errorMsg="Error Message"
        options={options}
      />
      <InputSelect
        id="id"
        name="name"
        hasError
        errorMsg="Error Message"
        options={options}
        placeholder="Placeholder"
      />
      <InputSelect
        id="id"
        name="name"
        label="With Label"
        hasError
        errorMsg="Error Message"
        options={options}
      />
      <InputSelect
        id="id"
        name="name"
        hint="With Hint"
        hasError
        errorMsg="Error Message"
        options={options}
      />
      <InputSelect
        id="id"
        name="name"
        label="With Label and Hint"
        hint="With Label and Hint"
        hasError
        errorMsg="Error Message"
        options={options}
      />
    </>
  );
};
