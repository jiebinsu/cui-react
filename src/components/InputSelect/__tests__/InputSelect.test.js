import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputSelect from "../index";

describe("InputSelect", () => {
  const options = [
    {
      value: "1",
      text: "one",
    },
    {
      value: "2",
      text: "two",
    },
  ];

  it("should render correctly", () => {
    const wrapper = mount(<InputSelect id="test" name="test" />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer
      .create(<InputSelect id="test" name="test" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render props correctly", () => {
    const tree = renderer
      .create(
        <InputSelect
          id="id"
          name="name"
          label="With Label and Hint"
          hint="With Label and Hint"
          hasError
          errorMsg="Error Message"
          options={options}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
