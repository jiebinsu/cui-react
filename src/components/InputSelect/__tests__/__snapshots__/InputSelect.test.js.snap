// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSelect should render correctly 1`] = `
<div
  className="coop-form__row"
>
  <select
    aria-describedby="test-hint"
    className="coop-form__field coop-form__select"
    id="test"
    name="test"
  />
</div>
`;

exports[`InputSelect should render props correctly 1`] = `
<div
  className="coop-form__row"
>
  <label
    className="coop-form__label"
    htmlFor="id"
  >
    With Label and Hint
  </label>
  <p
    className="coop-form__hint"
    id="id-hint"
  >
    This is an example hint
  </p>
  <p
    className="coop-form__error"
    id="id-error"
  >
    Error Message
  </p>
  <select
    aria-describedby="id-hint"
    className="coop-form__field coop-form__select coop-form__invalid"
    id="id"
    name="name"
  >
    <option
      value="1"
    >
      one
    </option>
    <option
      value="2"
    >
      two
    </option>
  </select>
</div>
`;
